### DISCUSSION TITLE
Detecting cancellation in Azure Pipelines

### DISCUSSION THREAD

**Discussion by mandalorianbob** [cake-build/cake#3616] [Q&A] [answered]
When building Unity, there's code I need to execute to return the license no matter what happens. I'm using Frosting.

Now, I want to make sure that the code executes when the build is cancelled on Azure Pipeline. 

I'm trying to figure out when/how the various Teardowns occur. I had the code in the "Finally" of the Frosting Task, but that didn't work.

However, I'm having a very difficult time testing this out. 

Does anyone know what Azure Pipelines sends when the build is cancelled, and whether or not Cake goes through the Task/Context teardowns when that happens?

**Comment by pascalberger** [comment]
You can use the [Cake.AzureDevOps addin](https://cakebuild.net/extensions/cake-azuredevops/) which provides aliases to retrieve information about an Azure Pipelines build, including the status and result: https://cakebuild.net/api/Cake.AzureDevOps.Pipelines/AzureDevOpsBuild/

**Reply by mandalorianbob** [reply]
Well that's good to know. I didn't know the DevOps addin had that information available to it. Unfortunately, I don't think that will help, as I meant during the build as it's happening. 

I can't run code if the build isn't running because it was cancelled. This isn't for like, detecting if some other build was cancelled, it's for detecting that a cancellation was just sent to the VM from Azure Pipelines.

I know there are some circumstances where a failing build will actually still trigger Teardown - for instance, when an Exception is thrown, Teardown still happens.

Do I need to manually hook into the the Console Application CancelKeyPress (which I ... _think_ is what Pipelines is using?) or is there something built into Cake that handles when the BuildSystem is trying to cancel the app?

**Comment by pascalberger** [comment, marked as answer]
Depending on what exactly you want to do, an option could be to add a second step to the Azure Pipeline with a [condition](https://docs.microsoft.com/en-us/azure/devops/pipelines/process/conditions?view=azure-devops&tabs=yaml) of `always()`, so that it is also called when pipeline is canceled and in this step you can call a specific target of the Cake build

**Reply by mandalorianbob** [reply]
Yeah that's an option, I was just hoping to keep everything within Cake, instead of having to add more logic to the Azure Pipelines. 

I'll try handling CancelKeyPress and update if that works in Azure Pipelines, otherwise, yeah I guess I'll have to handle it within Pipelines.

**Reply by mandalorianbob** [reply]
This ended up being the only solution that worked reliably.

**Reply by augustoproiete** [reply]
Interesting. Do you mean you added an event handler to the `CancelKeyPress` to detect the cancellation?

**Reply by mandalorianbob** [reply]
Sorry no, I meant the answer he gave was the one I went with. I put that logic into the Azure YAML pipeline with the condition of cancelled().

I tried the CancelKeyPress event and it did actually capture when the Agent sent CTRL-C to the app. Unfortunately when the Azure Agent sends that cancellation event, it only provides a 7 second timeout to respond, and that was apparently insufficient for Cake to run teardown. 

After getting it to work reliably locally, it failed on the very first Azure run I tried to cancel haha. So yeah, a YAML Task with condition cancelled() (or always()) is the only way I could reliably run code after cancellation. I was at least able to put the code in Cake and run it. (Where it took way less than 7 seconds, so I'm unsure why it failed.)
